### PR TITLE
mgr/cephadm: add count-per-host to PlacementSpec, and add support to scheduler

### PR DESCRIPTION
**Pull request by liewegas**
- for now, does not consider ports.. that is next
- refactors and simplifies the scheduling code to use fewer helpers and be more explicit
- drop the "avoid even numbers of monitors" logic that is IMO unnecessary and complicated

The next step is to handle port assignments...

Fixes: https://tracker.ceph.com/issues/47145

### CODE REVIEW

**Review by sebastian-philipp — APPROVED** [review]
nothing in here that raises red flags.

File: src/pybind/mgr/cephadm/schedule.py [hunk 1]
```diff
@@ -67,6 +68,7 @@ def __init__(self,
         self.filter_new_host = filter_new_host
         self.service_name = spec.service_name()
         self.daemons = daemons
+        self.allow_colo = allow_colo
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/pybind/mgr/cephadm/schedule.py:29]: there's not ``self.mgr`` here AFAICS

File: src/pybind/mgr/cephadm/schedule.py [hunk 2]
```diff
@@ -221,26 +198,37 @@ def remove_daemon_hosts(self, host_pool: List[HostPlacementSpec]) -> Set[DaemonD
...
-            return [
+        elif self.spec.placement.host_pattern:
+            hosts = [
                 HostPlacementSpec(x, '', '')
                 for x in self.spec.placement.filter_matching_hostspecs(self.hosts)
             ]
         # If none of the above and also no <count>
-        if self.spec.placement.count is None:
+        elif self.spec.placement.count is not None:
+            # backward compatibility: consider an empty placements to be the same pattern = *
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/pybind/mgr/cephadm/schedule.py]: fixed, along iwth bare ``count-per-host:N`` handling

**Review by jmolmo — APPROVED** [review]
It is needed to document the new "count-per-host" attribute in the [placement specification](https://docs.ceph.com/en/latest/cephadm/service-management/#placement-specification) point

**Comment by sebastian-philipp** [comment]
I think this is the cause of

https://pulpito.ceph.com/sage-2021-03-11_01:59:56-rados:cephadm-wip-sage-testing-2021-03-10-1754-distro-basic-smithi/5954929/


```
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/cephadm/serve.py", line 461, in _apply_all_services
    if self._apply_service(spec):
  File "/usr/share/ceph/mgr/cephadm/serve.py", line 595, in _apply_service
    forcename=slot.name)
  File "/usr/share/ceph/mgr/cephadm/module.py", line 548, in get_unique_name
    f'name {daemon_type}.{forcename} already in use')
orchestrator._interface.OrchestratorValidationError: name mgr.x already in use

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-11_07:16:51-rados-wip-kefu-testing-2021-03-11-1238-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-03-11_16:25:48-rados-wip-kefu-testing-2021-03-11-2158-distro-basic-gibba/
- https://pulpito.ceph.com/kchai-2021-03-12_05:42:44-rados-wip-kefu-testing-2021-03-12-1106-distro-basic-smithi/

**Pull request merged**